Corona-Warn-App/src/main/java/de/rki/coronawarnapp/risk/DefaultRiskLevels.kt

```diff
@@ -0,0 +1,246 @@
+import de.rki.coronawarnapp.R
+import de.rki.coronawarnapp.exception.RiskLevelCalculationException
+import de.rki.coronawarnapp.nearby.InternalExposureNotificationClient
+import de.rki.coronawarnapp.notification.NotificationHelper
+import de.rki.coronawarnapp.risk.RiskLevel.UNKNOWN_RISK_INITIAL
+import de.rki.coronawarnapp.risk.RiskLevel.UNKNOWN_RISK_OUTDATED_RESULTS
+import de.rki.coronawarnapp.server.protocols.ApplicationConfigurationOuterClass
+import de.rki.coronawarnapp.service.applicationconfiguration.ApplicationConfigurationService
+import de.rki.coronawarnapp.storage.LocalData
+import de.rki.coronawarnapp.storage.RiskLevelRepository
+import de.rki.coronawarnapp.util.TimeAndDateExtensions.millisecondsToHours
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import timber.log.Timber
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlin.math.round
+
+@Singleton
+class DefaultRiskLevels @Inject constructor() : RiskLevels {
+
+    companion object {
+
+        private var TAG = DefaultRiskLevels::class.simpleName
+
+        private const val DECIMAL_MULTIPLIER = 100
+    }
+
+    private fun calculateRiskScore(
+        attenuationParameters: ApplicationConfigurationOuterClass.AttenuationDuration,
+        exposureSummary: ExposureSummary
+    ): Double {
+
+        /** all attenuation values are capped to [TimeVariables.MAX_ATTENUATION_DURATION] */
+        val weightedAttenuationLow =
+            attenuationParameters.weights.low
+                .times(exposureSummary.attenuationDurationsInMinutes[0].capped())
+        val weightedAttenuationMid =
+            attenuationParameters.weights.mid
+                .times(exposureSummary.attenuationDurationsInMinutes[1].capped())
+        val weightedAttenuationHigh =
+            attenuationParameters.weights.high
+                .times(exposureSummary.attenuationDurationsInMinutes[2].capped())
+
+        val maximumRiskScore = exposureSummary.maximumRiskScore.toDouble()
+
+        val defaultBucketOffset = attenuationParameters.defaultBucketOffset.toDouble()
+        val normalizationDivisor = attenuationParameters.riskScoreNormalizationDivisor.toDouble()
+
+        val attenuationStrings =
+            "Weighted Attenuation: ($weightedAttenuationLow + $weightedAttenuationMid + " +
+                "$weightedAttenuationHigh + $defaultBucketOffset)"
+        Timber.v(attenuationStrings)
+
+        val weightedAttenuationDuration =
+            weightedAttenuationLow
+                .plus(weightedAttenuationMid)
+                .plus(weightedAttenuationHigh)
+                .plus(defaultBucketOffset)
+
+        Timber.v("Formula used: ($maximumRiskScore / $normalizationDivisor) * $weightedAttenuationDuration")
+
+        val riskScore = (maximumRiskScore / normalizationDivisor) * weightedAttenuationDuration
+
+        return round(riskScore.times(DECIMAL_MULTIPLIER)).div(DECIMAL_MULTIPLIER)
+    }
+
+    private fun Int.capped() =
```

Comment:
personally, I prefer to have override functions first, then public, then private